### ISSUE TITLE
Worst docs ever - wasted 8 hours and no luck!

### ISSUE THREAD

**Issue by alvipeo** [MicrosoftDocs/sql-docs#5796]
I've been trying to make a simple thing to work - to query Azure SQL table with Always Encrypted column (Azure Key Vault), all using this "mixed bag" docs page! All using .Net Core 5.0 web app which is deployed to App Service with Managed Identity.

First, Managed Identity works as I see all other queries working.

But when I query the table with Always Encrypted column I get this:
>Failed to decrypt a column encryption key. Invalid key store provider name: 'AZURE_KEY_VAULT'. A key store provider name must denote either a system key store provider or a registered custom key store provider. Valid system key store provider names are: 'MSSQL_CERTIFICATE_STORE', 'MSSQL_CNG_STORE', 'MSSQL_CSP_PROVIDER'. Valid (currently registered) custom key store provider names are: .

And my key vault access policies are correct.
Connection string does contain `Column Encryption Setting=enabled;`.

What's a big puzzle is how to make this all work. On this doc page you mixed always on and enclaves, .Net Framework and .Net Core. Why wouldn't you mix all possible languages here - Python, Java, etc?

Can we just have **full .Net Core 5.0 example** which would have everything together?

Here's the code I use for Managed Identity:
```
    internal static class AzureSqlConnection
    {
        internal static async Task<SqlConnection> GetSqlConnection(string connectionString, ILogger logger)
        {
            try
            {
                SqlConnection conn = new SqlConnection(connectionString);
                /*
                         * Always Encrypted with Managed Identity
                         *  https://docs.microsoft.com/en-us/azure/app-service/app-service-web-tutorial-connect-msi#modify-aspnet-core
                         *
                         */
#if !DEBUG
                conn.AccessToken = await new AzureServiceTokenProvider().GetAccessTokenAsync("https://database.windows.net/");
                logger.LogInformation($"token: {conn.AccessToken}");
#endif
                await conn.OpenAsync();
                return conn;
            }
            catch (Exception ex)
            {
                logger.LogError(ex, "Could not establish a connection to SQL Server");
                throw;
            }
        }
    }
}
```

But how do I enable Always Encrypted with Azure Key Vault?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 0ec4feb3-b3e1-67c9-0205-a93e78cfc821
* Version Independent ID: eb0e3b08-8d37-fc31-b011-a312bc2bbec0
* Content: [Using Always Encrypted with SqlClient - SQL Server](https://docs.microsoft.com/en-us/sql/connect/ado-net/sql/sqlclient-support-always-encrypted?view=sql-server-ver15)
* Content Source: [docs/connect/ado-net/sql/sqlclient-support-always-encrypted.md](https://github.com/MicrosoftDocs/sql-docs/blob/live/docs/connect/ado-net/sql/sqlclient-support-always-encrypted.md)
* Product: **sql**
* Technology: **connectivity**
* GitHub Login: @cheenamalhotra
* Microsoft Alias: **v-chmalh**

**Comment by alvipeo** [comment]
Here's my question on SO - [https://stackoverflow.com/q/65245840/2896495](https://stackoverflow.com/q/65245840/2896495).

**Comment by WilliamAntonRohm** [comment]
@alvipeo -- Alex, thank you for your feedback and question.  Please consider these resources:
- [SQL Server on Microsoft Q&A](https://docs.microsoft.com/en-us/answers/products/sql-server)
- [DBA Stack Exchange](https://dba.stackexchange.com/questions/tagged/sql-server)

@cheenamalhotra -- Cheena, please look into this issue.

cc @karinazhou for awareness on related issue #5799

**Comment by alvipeo** [comment]
In this article you mixed everything together. It's hard to follow and hard to apply to different cases. Please divide it - enclaves and key vaults, MSI and other auth. Please follow KISS principle. Same for examples.

**Comment by alvipeo** [comment]
Also, .NET Core 3.1 and .NET Core 5.0 (if applicable). Clearly state which version is supported for which scenarios.

**Comment by karinazhou** [comment]
Hi @alvipeo ,

I am sorry that this docs page doesn't help you to go through the issue you are facing. This docs page is for the general .NET application. Our samples are tested with .NET framework and .NET core console applications. I am afraid that we don't specific examples for ASP.NET application.

To explain it a bit, this docs page is always encrypted related which applies to all valid authentications that we support. For the authentication part, you can find information in [SQL Server Security](https://docs.microsoft.com/en-us/sql/connect/ado-net/sql/sql-server-security?view=sql-server-ver15) section which includes regular authentications, such as Windows Integrated auth and username-password auth, and also Active Directory authentications. 

The `Enabling Always Encrypted with secure enclaves` section is the enhanced version of the Always Encrypted feature. It works in a similar way as the first generation of Always Encrypted. So we include it in the same Always Encrypted documentation. In addition, the key store provider is necessary when using Always Encrypted columns. Azure Key Vault provider is one of those key store providers we support so that is the reason we have it on this page too. 

For the required .net platforms, you can find the information in the [Prerequisites ](https://docs.microsoft.com/en-us/sql/connect/ado-net/sql/sqlclient-support-always-encrypted?view=sql-server-ver15#prerequisites)section. There is a table illustrating the target frameworks supporting Always Encrypted. 

We appreciate your feedback and we will try to see how we can improve our document in the future. 


Thanks,

**Comment by karinazhou** [comment]
cc @David-Engel

**Comment by alvipeo** [comment]
I don't care is this is ASP.NET or console app. But I believe MSI has a lot of use in Azure, so is Azure Key Vault provider. And I wouldn't be wrong is they're used A LOT together. Yes, MSI is basically Azure AD but it's a special case and as I said it should be used a lot.

So I see here couple of *complete* examples:
1) console app (web app) which uses Azure AD (registered app with secret, etc) + Azure Key Vault as a key store provider
2) console/web app with MSI + Azure Key Vault provider
3) make it in .Net 3.1 **AND** .Net 5.0 (as you claim it should work for .Net Standard 2.1)

Put everything on GitHub so people could download it and "try it with Azure".
Probably do the same with Enclaves.

This will make sure that **every** single detail is covered - starting from Connection String and dependencies to MSI set up and Azure Key Vault policies, etc. This topic is NOT that easy so you can go away with 2 paragraphs about Key Vault.

**Comment by David-Engel** [comment]
@alvipeo Unfortunately we cannot write and maintain sample code for every single combination of ways developers want to use different components.

One concept I noticed that it appears you have missed in your original post is that Azure Key Vault requires its own authentication. You have only implemented MSI for the database authentication. Since the premise of Always Encrypted is that the client does not trust the database, the master key needed for decryption is stored entirely separately from any type of database authentication. Thus you need to configure your chosen key store provider (Azure Key Vault in your case). Since Azure Key Vault is cloud-based (compared to local key stores like the Windows certificate store, which already know your local application security context), you need to provide a way to authenticate.

If you start with the [Using Azure Key Vault with Always Encrypted](https://docs.microsoft.com/en-us/sql/connect/ado-net/sql/azure-key-vault-example?view=sql-server-ver15) example, you simply need to modify the AzureActiveDirectoryAuthenticationCallback method to obtain an access token using MSI instead of clientId/secret in the example. I suspect you can simple replace the entire contents of that method with something almost as simple as another call to AzureServiceTokenProvider.

Regards,
David

**Comment by alvipeo** [comment]
@David-Engel We neither.

I understand the difference between MSI and db auth and Key Vault auth. I tried that example and it didn't work with .Net 5.0. I'll try with .Net Core 3.1.

**Comment by cheenamalhotra** [comment]
Issue is followed in https://github.com/dotnet/SqlClient/issues/847, closing from here.
